src/target/riscv/debug_register_printers.c

```diff
@@ -0,0 +1,79 @@
+#include <stdio.h>
+#include <inttypes.h>
+#include <assert.h>
+#include <stdarg.h>
+
+#include "debug_register_printers.h"
+
+unsigned int get_len_or_sprintf(char *buf, unsigned int curr, const char *format, ...)
+{
+	assert(format);
+	va_list args;
+	int length;
+
+	va_start(args, format);
+	if (buf)
+		length = vsprintf(buf + curr, format, args);
+	else
+		length = vsnprintf(NULL, 0, format, args);
+	va_end(args);
+	assert(length >= 0);
+	return (unsigned int)length;
+}
+
+unsigned int print_value(char *buf, unsigned int curr, uint64_t value)
+{
+	const char * const format = value > 9 ? "0x%" PRIx64 : "%" PRIx64;
+
+	return get_len_or_sprintf(buf, curr, format, value);
+}
+
+unsigned int riscv_debug_reg_field_value_to_s(char *buf, unsigned int curr,
+		enum riscv_debug_reg_ordinal reg_ordinal,
+		struct riscv_debug_reg_ctx context,
+		unsigned int offset, uint64_t field_value)
+{
+	const char * const field_value_name = riscv_debug_reg_decode_field(
+			reg_ordinal, context, offset, field_value);
+
+	if (!field_value_name)
+		return print_value(buf, curr, field_value);
+	return get_len_or_sprintf(buf, curr, "%s", field_value_name);
+}
+
+unsigned int riscv_debug_reg_field_to_s(char *buf, unsigned int curr,
+		enum riscv_debug_reg_ordinal reg_ordinal,
+		struct riscv_debug_reg_ctx context,
+		unsigned int offset, uint64_t field_value)
+{
+	const char * const field_name = riscv_debug_reg_field_name(reg_ordinal,
+			context, offset);
+	assert(field_name);
+	const unsigned int name_len = get_len_or_sprintf(buf, curr, "%s=", field_name);
+
+	return name_len + riscv_debug_reg_field_value_to_s(buf, curr + name_len,
+			reg_ordinal, context, offset, field_value);
+}
+
+unsigned int riscv_debug_reg_to_s(char *buf, enum riscv_debug_reg_ordinal reg_ordinal,
+		struct riscv_debug_reg_ctx context, uint64_t value)
+{
+	unsigned int length = 0;
+
+	length += print_value(buf, length, value);
+	length += get_len_or_sprintf(buf, length, " { ");
+
+	const unsigned int width = riscv_debug_reg_width(reg_ordinal, context);
+
+	for (unsigned int offset = 0; offset < width; ++offset) {
+		if (!riscv_debug_reg_field_is_present(reg_ordinal, context, offset))
+			continue;
```

Comment:
This looks really strange to me. You're iterating over every bit, and then you call a function for every bit offset to check if a field starts there, just to visit every field. Instead, it seems much better to have a list of fields for each register and then you can iterate over that.

Something like:
```
struct reg_field_info {
   const char *name;
   unsigned int lsb; // inclusive
   unsigned int msb; // inclusive
   const char **values; // If non-NULL, array of human-readable string for each possible value
};

struct reg_info {
   const char *name;
   struct reg_field_info fields[];
};
```

Reply:
Addressed.